### DISCUSSION TITLE
v2.0 Ideas

### DISCUSSION THREAD

**Discussion by AnidemDex** [dialogic-godot/dialogic#751] [Ideas]
_Hey, remember me? The guy that tried to rebase the main branch with resources? I'm back_

So, while waiting for 1.4, I want to list the things that can be added for 2.0, including those that can potentially break the whole dialogic scheme, so all ideas are welcome. and they can be useful for Emilio.

For now, I can think on splitting the whole dialogic plugin behaviour to make it easier to maintain to current and new maintainers, for anyone that wanted to contribute and add/improve the functionality of the plugin _(yes, is possible, see [this dialog node](https://github.com/AnidemDex/Godot-DialogPlugin) and this [sequential event manager](https://github.com/AnidemDex/Godot-EventSystem))_

## Ideas list
- [x] Dialogic node would be a proper node that only takes care about displaying dialogs (https://github.com/coppolaemilio/dialogic/issues/692 and https://github.com/coppolaemilio/dialogic/issues/276 can take advantage of this)
- [x] The processing of the dialogic events would be take care of another class, and not to be hardcoded in dialogic node
- [x] Dialogic theme editor can use the integrated `Theme` class that Godot has with some flavour, instead of trying to recreate the whole wheel.
- [x] Dialogic definitions can use variables from objects, instead of trying to recreate those variables in a collection.
- [x] https://github.com/coppolaemilio/dialogic/issues/580
- [x] Text speed should be modified through the node, not through themes
- [ ] Add a "recent used characters" selector
- [ ] Adding an image into the viewed scene that doesn't behave the same way as a character, but also isn't a background object and have to be part of the background image
- [x] https://github.com/coppolaemilio/dialogic/issues/608

---
Next points comes from https://github.com/coppolaemilio/dialogic/issues/884
- [ ] **Add a "Y offset value" in the "update character node"** that let us move the player besides the predetermined position
- [ ] Makes the "Join character node" to add multiple characters at a time with only one node. (Alternative, a character event join that adds multiple characters)
- [ ]  Change the way Dialogic show where a node is going to be put. 
---

The first two options are something possible and easy to implement, you can see https://github.com/AnidemDex/Godot-DialogPlugin/pull/56 where the plugin has two separable parts. In Dialogic, those two parts can be combined in one by default, bu they don't interfer with each other.

Hey, I've openned a discussion for integrating a dialog node here, in case you want to share your thoughts about it https://github.com/coppolaemilio/dialogic/discussions/832

**Comment by Jowan-Spooner** [comment]
Related:
- #580
- #692
- #276

I'm not really sure what the best steps are yet, but here are the (very general) goals I could think of for 2.0:
- The dialogic interface as a central part of what makes dialogic unique should work relatively similar (from a user perspective).
- The possibility of integrating the dialog as easily as now (drag and drop/two lines of code) should be kept.
I agree that it would improve the usability of the plugin, if more things became optional. For me, here are a few ways, I could see dialogic used:

- to create a VISUAL NOVEL by someone who wants to code as little as possible (or at least some people in the team don't want to). For these people the editor and the DialogNode should do as much of the background stuff as possible.
- only as an TEXT/STORY EDITOR, but executing it on their own. I think the Dialogic editor is really nice and I'd personally like to use it and then work with the resulting stuff myself (without necessarily having to understand id's and all the setting names
- only as a DIALOG EXECUTER. Surprisingly many people want to create the majority of the dialog procedurally, which would be kinda cool to allow as well, although it is quite far from what Dialogic was created for. There are also people who would like to write stories in different formats (see 580) but want dialogic to display them. 
- something in between. I personally would like if it was possible to only use parts of the DialogNode. E.g. have it manage portraits but allowing me to do themeing myself, or vice versa. 

I'm just trying to think of what dialogic is for, to find out what the best way of providing to all these needs would be. 

Some changes I would like to see:
- Generally better API for everything. Examples: custom portraits reacting to the DialogNode, changing settings through code, etc.
- Improving data storage (looking at all sorts of strange dictionaries here 😬)
- Improving data evaluation (looking at kilometres of if statements here 😩)

In the end I think all changes must either:
- improve something for the user (e.g. a new feature, a better workflow)
- improve the process of adding new features for us developers

I'm not a fan of making things more complicated with dialogic 2.0 becoming less usefull in the end.

Have to think about all of this a bit more though. Good night.

**Reply by AnidemDex** [reply]
updated

**Comment by coppolaemilio** [comment]
> Dialogic theme editor can use the integrated Theme class that Godot has with some flavour, instead of trying to recreate the whole wheel.

I don't really like the Godot's built in theme editor. Even if it was updated recently and it is much better, there are so many controls it makes it really hard to explain to non-technical people which are the target audience of this plugin. There is no reason why an average visual novel writer should learn the draconian rules of the overwriting theme parameters by clicking on "manage items" importing and navigating all those confusing menus. I'm not discarding the idea of allowing for native themes to be used but they should not be default solution for people using the plugin.

> Dialogic definitions can use variables from objects, instead of trying to recreate those variables in a collection.

Yes, using variables from scenes should be the right way to handle this.

**Reply by AnidemDex** [reply]
The idea of using Theme resource is not to make users learn about it, but more to make easier for them in case they wanted to make their own theme through code, or apply the dialogic theme to a node that they made. The current implementation of themes in dialogic can be there and the user will not notice that through the dialogic editor

**Comment by AnidemDex** [comment]
Now that I remember, I suggest to use a different approach to get things on the tree. The `get_parent().get_parent()...` hell and the longs `$Very/very/very/long/node/nested/path` tends to become a nightmare if you want to make a change

**Comment by Sslaxx** [comment]
Unsure of how useful it would be, but maybe take a look at the way https://github.com/nathanhoad/godot_dialogue_manager handles functions and variables (as this sounds similar to what you're proposing)?

**Comment by ii4y-studios** [comment]
Text speed should be an exposed master-level variable that can be user set instead of per theme, so unless some bbcode for an individual box changes the speed overwrites it the game uses the master level speed. I know a lot of people have different preferences for text speeds (I prefer the fastest I can get, immediate if available, other people I know prefer varying levels of slower) so it should be user-selectable

**Reply by AnidemDex** [reply]
updated

**Comment by ii4y-studios** [comment]
One thing that would be nice if possible would be to have a like short list on the character selector for recently used. With say the last five used characters in the session, to make selecting them faster when you're working on a scene and putting in a ton of text nodes, where you're probably going to be using mostly the same characters for one particular scene. So like a tree like this, with the 'All Characters' then going into the normal tree as it shows currently:

---------
Alice
Bob
Charlie 
David
Erin
All Characters -> 
---

Or another thought would be instead of just the last 5 selected chronologically, having it like smartly show any that have been joined into a scene from any Character Joins in that particular timeline, with again the option to select All Characters at the end of the short list. 

Either way, sometimes currently I end up just copying and pasting text nodes to preserve a character instead of having to go through the big tree to select them again which isnt convenient as you have more complex character numbers and layouts.

**Reply by AnidemDex** [reply]
updated

**Comment by ii4y-studios** [comment]
One that just popped into my head and I might forget later, an option to set on a character to ignore the Portrait Dim settings if rheyre being used. My idea is for something I see in VNs sometimes, where like instead of (or addition to) characters in a scene a non-character item is shown inset over top the background.  And you could do thst with a character but you'd have to change the theme, and if you're using fade overall then while the themes changed you wouldn't have it.

Or, if not that, then just some other kind of built-in action to add some image as a second layer above the background layer

**Reply by AnidemDex** [reply]
Maybe I readed it wrong, but this sounds like many points in one

**Reply by ii4y-studios** [reply]
its two alternate implementation ideas for the same one idea: adding an image into the viewed scene that doesn't behave the same way as a character, but also isn't a background object and have to be part of the background image

**Reply by AnidemDex** [reply]
updated

**Comment by ii4y-studios** [comment]
I'm gonna add this here, don't think it necessarily needs to be 2.0 if its not gonna be easy to do, but as a 2.x general feature. I want to be able to call timelines as subroutines, insteaed of gotos. 

for more solid example of what i mean, I've got a number of small discussions that *could* happen in any number of different places with multiple entry points. being able to make that just one timeline, instead of multiple blocks in each timeline that it can possible happen in it, would make it easier to maintain for any changes to the wording (or translations) that need to happen by having it in only one place. so in all of those places I have a variable check to see if it has alraedy happened yet, if not then gosub to that discussion, set that it's been completed, and return back to the timeline it branched off of to continue where it was from there

**Reply by Jowan-Spooner** [reply]
Maybe as like a special mode for the Jump (the new change timeline) event? So that it will go back to the jump event if the timeline it went to ends?

**Reply by ii4y-studios** [reply]
yeah, that way would work, it just needs to return to where it left off on the original timeline

**Comment by M7mdKady14** [comment]
Not sure if the has been posted before but what about timed choices?
for example you have four buttons for four choices but you can enable a "timer" functionality on any of them that does a number of things when the timer runs out, for example disabling the option to choose that choice or the game chooses a specific or random choice.

**Reply by coppolaemilio** [reply]
With the new display nodes system this is already possible but there isn't any "native" way of doing it. If it gets added to the core of the plugin it will most likely be for version 2.1
